include/execpools/thread_pool_base.hpp

```diff
@@ -0,0 +1,485 @@
+      template <stdexec::sender Sender, std::integral Shape, class Fun>
+      using bulk_sender_t =
+        stdexec::__t<bulk_sender<stdexec::__id<stdexec::__decay_t<Sender>>, Shape, Fun>>;
+
+      STDEXEC_MEMFN_FRIEND(bulk);
+
+      template <stdexec::sender S, std::integral Shape, class Fun>
+        STDEXEC_MEMFN_DECL(auto bulk)(this const scheduler& sch, S&& sndr, Shape shape, Fun fun) noexcept
+        -> bulk_sender_t<S, Shape, Fun> {
+        return bulk_sender_t<S, Shape, Fun>{
+          *sch.pool_, static_cast<S&&>(sndr), shape, static_cast<Fun&&>(fun)};
+      }
+
+      [[nodiscard]]
+      constexpr auto
+        forward_progress_guarantee() const noexcept -> stdexec::forward_progress_guarantee {
+        return pool_->forward_progress_guarantee();
+      }
+
+      friend thread_pool_base;
+
+      explicit scheduler(DerivedPoolType& pool) noexcept
+        : pool_(&pool) {
+      }
+
+      DerivedPoolType* pool_;
+
+     public:
+      using __t = scheduler;
+      using __id = scheduler;
+      auto operator==(const scheduler&) const -> bool = default;
+
+      constexpr auto query(stdexec::get_forward_progress_guarantee_t) const noexcept
+        -> stdexec::forward_progress_guarantee {
+        return forward_progress_guarantee();
+      }
+
+      auto schedule() const noexcept -> sender {
+        return sender{*pool_};
+      }
+    };
+
+    [[nodiscard]]
+    auto get_scheduler() noexcept -> scheduler {
+      return scheduler{static_cast<DerivedPoolType&>(*this)};
+    }
+
+    /* Is this even needed? Looks like no.
+     * void request_stop() noexcept {
+        // Should this do anything? TBB supports its own flavor of cancelation at a tbb::task_group level
+        //
+    https://spec.oneapi.io/versions/latest/elements/oneTBB/source/task_scheduler/scheduling_controls/task_group_context_cls.html
+        // but not at the tbb::task_arena level.
+        // https://spec.oneapi.io/versions/latest/elements/oneTBB/source/task_scheduler/task_arena/task_arena_cls.html
+    }*/
+
+    [[nodiscard]]
+    auto available_parallelism() const -> std::uint32_t {
+      return static_cast<DerivedPoolType&>(*this).available_parallelism();
+    }
+
+   private:
+    void enqueue(task_base* task, std::uint32_t tid = 0) noexcept {
+      static_cast<DerivedPoolType&>(*this).enqueue(task, tid);
+    }
+
+    void bulk_enqueue(task_base* task, std::uint32_t n_threads) noexcept {
+      for (std::uint32_t tid = 0; tid < n_threads; ++tid) {
+        this->enqueue(task, tid);
+      }
+    }
+  };
+
+  template <class PoolType, class ReceiverId>
+  struct operation<PoolType, ReceiverId>::__t : task_base {
+    using __id = operation;
+    friend class thread_pool_base<PoolType>;
+
+    PoolType& pool_;
+    Receiver rcvr_;
+
+    explicit __t(PoolType& pool, Receiver rcvr)
+      : pool_(pool)
+      , rcvr_(std::move(rcvr)) {
+      this->__execute =
+        [](task_base* t, std::uint32_t /* tid What is this needed for? */) noexcept {
```

Comment:
I never got an answer to what the uint32 is for.

Reply:
When I checked the code I was also wondering. For me it looks like used during bulk tasks this is used to propagate/report the error in the proper into the proper thread. Thus, I thought I'm keeping this comment here, because I'm not 100% sure.